### PR TITLE
Implement Legacy RegExp features

### PR DESCRIPTION
**Pull request by kisbg**
Signed-off-by: bence gabor kis <kisbg@inf.u-szeged.hu>

### CODE REVIEW

**Comment by clover2123** [comment]
I cannot find `Legacy RegExp` features in the current [draft ECMA-262](https://tc39.es/ecma262/).
It seems that this spec is in the proposal stage and not yet written on the standard.
How about leaving this pr on pending until it is updated?

**Comment by kisbg** [comment]
@clover2123  Sorry for the late reply! This pull request based on this proposal documentation `https://github.com/tc39/proposal-regexp-legacy-features`.

Of course i can close this pull request and wait for a more official standard documentation.

**Comment by clover2123** [comment]
@kisbg I've updated test262 to the latest version (#834)
After #834 landed, would you please check and re-run relevant test cases in the new test262?

**Comment by kisbg** [comment]
@clover2123  Of course!

**Comment by kisbg** [comment]
Updated the pull request and added the related test-cases in test262

File: src/runtime/GlobalObjectBuiltinRegExp.cpp [hunk 1]
```diff
@@ -641,36 +647,79 @@ static Value builtinRegExpUnicodeGetter(ExecutionState& state, Value thisValue,
...
-    return state.resolveCallee()->codeBlock()->context()->regexpStatus().input;
+    return getLegacyRegExpStaticProperty(state, thisValue, state.resolveCallee()->codeBlock()->context()->regexpStatus().input);
 }
 
 static Value builtinRegExpInputSetter(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
 {
+    if (!thisValue.isPointerValue() || thisValue.asPointerValue() != state.context()->globalObject()->regexp()) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, "thisValue is not RegExp");
+    }
     state.resolveCallee()->codeBlock()->context()->regexpStatus().input = argv[0].toString(state);
```
Comment (clover2123) [inline src/runtime/GlobalObjectBuiltinRegExp.cpp:681]: Why don't you call `setLegacyRegExpStaticProperty` here?

**Review by clover2123 — CHANGES_REQUESTED** [review]

File: src/runtime/GlobalObjectBuiltinRegExp.cpp [hunk 2]
```diff
@@ -752,43 +801,47 @@ void GlobalObject::installRegExp(ExecutionState& state)
...
+        m_regexp->defineOwnProperty(state, strings->input, ObjectPropertyDescriptor(gs, (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::ConfigurablePresent)));
+        m_regexp->defineOwnProperty(state, strings->$_, ObjectPropertyDescriptor(gs, (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::ConfigurablePresent)));
     }
 
     {
         JSGetterSetter gs(
             new NativeFunctionObject(state, NativeFunctionInfo(strings->get, builtinRegExpLastMatchGetter, 0, NativeFunctionInfo::Strict)), Value(Value::EmptyValue));
-        m_regexp->defineOwnProperty(state, strings->lastMatch, ObjectPropertyDescriptor(gs, (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::EnumerablePresent)));
-        m_regexp->defineOwnProperty(state, strings->$Ampersand, ObjectPropertyDescriptor(gs, (ObjectPropertyDescriptor::PresentAttribute)(ObjectPropertyDescriptor::NotPresent)));
+        new NativeFunctionObject(state, NativeFunctionInfo(strings->set, builtinRegExpLastMatchSetter, 1, NativeFunctionInfo::Strict));
```
Comment (clover2123) [inline src/runtime/GlobalObjectBuiltinRegExp.cpp:811]: This newly allocated `NativeFunctionObject` seems useless.
And there are other similar cases like this. Please check and remove it.

File: src/runtime/GlobalObjectBuiltinRegExp.cpp [hunk 3]
```diff
@@ -641,36 +647,79 @@ static Value builtinRegExpUnicodeGetter(ExecutionState& state, Value thisValue,
...
 }
 
 static Value builtinRegExpLastMatchGetter(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
 {
-    return Value(new StringView(state.resolveCallee()->codeBlock()->context()->regexpStatus().lastMatch));
+    return getLegacyRegExpStaticProperty(state, thisValue, Value(new StringView(state.resolveCallee()->codeBlock()->context()->regexpStatus().lastMatch)));
+}
+
+static Value builtinRegExpLastMatchSetter(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
+{
```
Comment (clover2123) [inline src/runtime/GlobalObjectBuiltinRegExp.cpp:691]: According to the standard, `RegExp.lastMatch` does not have setter (getter only).
Remove this setter function and other setter functions too.

**Review by clover2123 — CHANGES_REQUESTED** [review]

**Comment by clover2123** [comment]
`GetLegacyRegExpStaticProperty` and `SetLegacyRegExpStaticProperty` are abstract operations that means we don't have to implement these functions exactly as spec.
How about simplify legacy property's getter/setter functions by using some macro instead of calling the above function in itself each time?

**Comment by kisbg** [comment]
@clover2123  I updated the pull request!

File: src/runtime/GlobalObjectBuiltinRegExp.cpp [hunk 4]
```diff
@@ -640,37 +646,50 @@ static Value builtinRegExpUnicodeGetter(ExecutionState& state, Value thisValue,
 {
     return builtinRegExpOptionGetterHelper(state, thisValue, RegExpObject::Option::Unicode);
 }
+#define DEFINE_GETTER_LEGACY_REGEXP_PROPERTY(stringView)                                                          \
+    if (!thisValue.isPointerValue() || thisValue.asPointerValue() != state.context()->globalObject()->regexp()) { \
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, "thisValue is not RegExp");                 \
+    }                                                                                                             \
+    if (stringView.isEmpty()) {                                                                                   \
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, "Value is empty");                          \
```
Comment (clover2123) [inline src/runtime/GlobalObjectBuiltinRegExp.cpp:654]: For error messages, how about exploiting existing messages defined in `ErrorObject::Messages`?
e.g) `ErrorObject::Messages::GlobalObject_ThisNotRegExpObject` for the first error

File: src/runtime/RegExpObject.cpp [hunk 5]
```diff
@@ -457,6 +458,18 @@ void RegExpObject::createRegexMatchResult(ExecutionState& state, String* str, Re
     } while (testResult);
 }
 
+static void invalidateLegacyRegExpStaticProperties(ExecutionState& state)
+{
+    state.resolveCallee()->codeBlock()->context()->regexpStatus().input = String::emptyString;
+    state.resolveCallee()->codeBlock()->context()->regexpStatus().lastMatch = StringView();
+    state.resolveCallee()->codeBlock()->context()->regexpStatus().lastParen = StringView();
+    state.resolveCallee()->codeBlock()->context()->regexpStatus().leftContext = StringView();
+    state.resolveCallee()->codeBlock()->context()->regexpStatus().rightContext = StringView();
```
Comment (clover2123) [inline src/runtime/RegExpObject.cpp:467]: It seems like that `RegExpStatus` could be accessed simply by `state.context()->regexpStatus()`

**Comment by kisbg** [comment]
@clover2123  Thank you for the review!

**Comment by clover2123** [comment]
It seems that you have reverted the previous update.
Please check my comments once more.

**Comment by kisbg** [comment]
Sorry for the inconvenience!

**Review by clover2123 — APPROVED** [review]
LGTM

**Review by ksh8281 — APPROVED** [review]

**Pull request merged**